smfile_writer.py

```diff
@@ -1,152 +1,152 @@
-    note_256        = round(measure_sec/256, 5)                                                    #time of notes in seconds
-
-    for i in range(measure_all):
-        measure = []
-        for j in timings:
-            if((i * measure_sec) <= j < ((i+1) * measure_sec)):
-                measure.append(round(j-(i*measure_sec), 5))
-        note.extend(calculate_measure(measure, note_256))
-    note[-1] = ';'
-    return note
+    # pre-generate output data
+    title  = '#TITLE:%s;\n' % step_dict['title']
+    artist = '#ARTIST:jhaco vs cpuguy96;\n'
+    music  = '#MUSIC:%s.ogg;\n' % file_name
+    select = 'SELECTABLE:YES;\n'
+    bpm    = 'BPMS:0.000=%s;\n\n' % str(step_dict['bpm'])
+    note_data = ''
+    for difficulty in step_dict['notes'].keys():
+        note_data += '//---------------dance-single - ----------------\n'
+        note_data += '#NOTES:\n'
+        note_data += '     dance-single:\n'
+        note_data += '     :\n'
+        note_data += '     %s:\n' % difficulty
+        note_data += '     8:\n'
+        note_data += '     1.000,1.000,1.000,1.000,1.000:\n'
+        for note in step_dict['notes'][difficulty]:
+                note_data += note + '\n'
+
+    with open(join(output_dir, ofile), 'w') as f:
+        f.write(''.join((title, artist, music, select, bpm, note_data))) # calling write only once is best practice
+        
+#===================================================================================================
+
+def find_gcd(note_positions):
+    x = reduce(gcd, note_positions + [256])
+    return x
+
+def generate_measure(notes, note_positions):
+    # tries to fit 256, 128, 64, 32, 16, 8 or 4 note measures
+    note_gcd = find_gcd(note_positions)
+
+    if note_gcd == 128: # if fitting returns a 2 note measure, adjust to 4 note
+        note_gcd = 64
+
+    # place notes in generated measure
+    generated_measure = ['0000']*int(256/note_gcd)
+    for note, p in enumerate(note_positions):
+        new_p = int(p/note_gcd)-1
+        generated_measure[new_p] = notes[note]
+
+    return generated_measure
+
+def notes_to_measure(notes, timings, note_256):
+    if not notes or not timings: # no notes in current measure
+        return ['0000','0000','0000','0000',',']
+    
+    note_positions = []
+    for time in timings:
+        note_position = 0
+        min_error = 5.0 # edge case: might need to adjust for really slow songs
+        for i in range(256):
+            error = abs(time - i*note_256)
+            if error <= min_error: # narrows down note position based on timing difference
+                note_position = i
+                min_error = error
+            else:
+                break # stops finding note position once found
+        note_positions.append(note_position+1) #pre-adjusts positions to be mathematically correct
+    measure = generate_measure(notes, note_positions)
+    measure.append(',')

```

Comment:
It will be better to do this outside of this function since it may not be desirable to always add this afterwards.